### ISSUE TITLE
[bug] Cannot write tests, undefined variables

### ISSUE THREAD

**Issue by ospfranco** [tauri-apps/tauri#8996]
### Describe the bug

I'm trying to get a basic test suite running in an app but keep hittings snags.

1. I'm using Bun, so already hitting some incompatibility issues. Not Tauri's fault, just here for documentation: jsdom is not compatible, so using happydom, not sure if this will bring problems down the line. However I could not find any official documentation on getting a test suite to run, had to rely on a [3rd party](https://yonatankra.com/how-to-setup-vitest-in-a-tauri-project/)

2. As soon as I sorted that out got a different error about missing __TAURI_IPC__, [this comment suggested bumping to beta version](https://github.com/tauri-apps/tauri/issues/3434#issuecomment-1943682845). That solved that specific undefined module, but now facing:

3. [undefined __TAURI_INTERNALS__](https://github.com/tauri-apps/tauri/issues/8369)

Am I going down the right rabbit hole in here? Are these issues known to the community?

### Reproduction

Here is the actual repo, you only need to check out the `osp/tests` branch and run `bun install` and `bun test`

### Full `tauri info` output

```text
[✔] Environment
    - OS: Mac OS 14.3.1 X64
    ✔ Xcode Command Line Tools: installed
    ✔ rustc: 1.73.0 (cc66ad468 2023-10-03)
    ✔ cargo: 1.73.0 (9c4383fb5 2023-08-26)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-aarch64-apple-darwin (default)
    - node: 18.19.0
    - yarn: 4.0.2
    - npm: 10.2.3
    - bun: 1.0.29

[-] Packages
    - tauri [RUST]: 1.5.4
    - tauri-build [RUST]: 1.5.1
    - wry [RUST]: 0.24.7
    - tao [RUST]: 0.16.7
    - @tauri-apps/api : not installed!
    - @tauri-apps/cli [NPM]: 1.5.10

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist
    - devPath: http://localhost:1420/
    - framework: React
    - bundler: Vite
```

**Comment by FabianLars** [comment]
1. happydom _should_ work but i guess it's not tested as much as jsdom so take that with a grain of salt. Both the guide you linked and your repo commit is missing the stuff this talks about https://tauri.app/v1/guides/testing/mocking. In an actual tauri app, the rust backend will set the window variables but since there's no rust part in the frontend testing you must mock them here.

2. The comment you linked to was from a completely different context and the issue there was just a version mismatch. Speaking of version mismatches, you updated your js packages to v2, and added the shell plugin for v2, but kept the tauri and tauri-build crates at 1.5 which will cause issues in the actual tauri app.

3. Same issue as 1) we just changed the name of the variable on the window object, but it's essentially the same.

tldr: mocking is required in frontend tests, see the link above. It also uses vitest.

**Comment by ospfranco** [comment]
Thanks for clarifying!

1. I thought it was possible to write the e2e tests without the need for webdriver, since it doesn't support macos it's a big annoyance. 

2.For now just tyring to get tests to work, but thanks for pointing out the crates issue. As for a different context, yeah, I know, I just got a bunch of undefined objects but there is no information on how to mock them or polyfill them.

3. How can this be mocked polyfilled then? No idea about the internals of Tauri and what needs to be there so I can run the tests.

**Comment by FabianLars** [comment]
> I thought it was possible to write the e2e tests without the need for webdriver, since it doesn't support macos it's a big annoyance.

Sadly not. vitest and similar things can only test the frontend/js part which is missing all the actual tauri specific stuff (because that's all in the rust side). So for e2e you need webdriver that is able to actually run the rust side and the webview too :/

> How can this be mocked polyfilled then? No idea about the internals of Tauri and what needs to be there so I can run the tests.

https://tauri.app/v1/guides/testing/mocking/ or the docs for the specific api: <https://tauri.app/v1/api/js/mocks#mockipc>
This will create the `window.__TAURI_IPC__` (or `__internals__` if you use v2) variable and is also used to actually mock the return values.

**Comment by ospfranco** [comment]
I understand there is a way to mock the native APIs, but there is no information what do the internal modules actually do, and what exactly needs to be mocked. I hoped to get a bit more practical information. Thanks anyway.
